core/src/main/java/io/grpc/internal/DelayedStream.java

```diff
@@ -56,263 +54,227 @@
 
-  /**
-   * Transfers all pending and future requests and mutations to the given stream.
-   *
-   * <p>No-op if either this method or {@link #cancel} have already been called.
-   */
-  final void setStream(ClientStream stream) {
-    synchronized (this) {
-      if (error != null || realStream != null) {
-        return;
-      }
-      realStream = checkNotNull(stream, "stream");
-      // listener can only be non-null if start has already been called.
-      if (listener != null) {
-        startStream();
+    delayOrExecute(new Runnable() {
+      @Override
+      public void run() {
+        realStream.start(listener);
       }
-    }
+    });
   }
 
   @Override
-  public void writeMessage(InputStream message) {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingMessages.add(new PendingMessage(message, messageCompressionEnabled));
-          return;
+  public void writeMessage(final InputStream message) {
+    checkNotNull(message, "message");
+    if (passThrough) {
+      realStream.writeMessage(message);
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.writeMessage(message);
         }
-      }
+      });
     }
-    startedRealStream.writeMessage(message);
   }
 
   @Override
   public void flush() {
-    if (startedRealStream == null) {
-      synchronized (this) {
-        if (startedRealStream == null) {
-          pendingFlush = true;
-          return;
+    if (passThrough) {
+      realStream.flush();
+    } else {
+      delayOrExecute(new Runnable() {
+        @Override
+        public void run() {
+          realStream.flush();
         }
-      }
+      });
     }
-    startedRealStream.flush();
   }
 
   @Override
   public void cancel(Status reason) {
-    // At least one of them is null.
-    ClientStream streamToBeCancelled = startedRealStream;
-    ClientStreamListener listenerToBeCalled = null;
-    if (streamToBeCancelled == null) {
-      synchronized (this) {
-        if (realStream != null) {
-          // realStream already set. Just cancel it.
-          streamToBeCancelled = realStream;
-        } else if (error == null) {
-          // Neither realStream and error are set. Will set the error and call the listener if
-          // it's set.
-          error = checkNotNull(reason);
-          realStream = NoopClientStream.INSTANCE;
-          if (listener != null) {
-            // call startStream anyways to drain pending messages.
-            startStream();
-            listenerToBeCalled = listener;
-          }
-        }  // else: error already set, do nothing.
+    checkNotNull(reason, "reason");
+    boolean cancelNow;
+    ClientStreamListener listenerToClose = null;
+    synchronized (this) {
+      cancelNow = passThrough;
+      if (!cancelNow) {
+        error = reason;
+        listenerToClose = listener;
+        // TODO(ejona): Close InputStreams?
```

Comment:
Currently `MessageFramer` doesn't. At present that doesn't hurt much, but we will need to fix places eventually.